### PR TITLE
AC-1077: Add/Enhance Observation APIs

### PR DESCRIPTION
**Pull request by shubhamsgit**
## Description of what I changed

- This PR is for the enhancement of the Observation APIs in the SDK

## Issue I worked on

JIRA Issue: https://issues.openmrs.org/browse/AC-1077

## Checklist: I completed these to help reviewers :)

- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [ ] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/EncounterRepository.kt [hunk 1]
```diff
@@ -3,11 +3,7 @@ package com.openmrs.android_sdk.library.api.repository
 import com.openmrs.android_sdk.library.dao.PatientDAO
 import com.openmrs.android_sdk.library.dao.VisitDAO
 import com.openmrs.android_sdk.library.databases.AppDatabaseHelper
-import com.openmrs.android_sdk.library.databases.entities.ConceptEntity
-import com.openmrs.android_sdk.library.models.Encounter
-import com.openmrs.android_sdk.library.models.EncounterType
-import com.openmrs.android_sdk.library.models.Encountercreate
-import com.openmrs.android_sdk.library.models.ResultType
+import com.openmrs.android_sdk.library.models.*
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/EncounterRepository.kt:6]: Avoid star imports

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+) : BaseRepository() {
+
+    /**
+     * Create an Observation record on the server from the locally
+     * stored observation entity record
+     *
+     * @param observationEntity the locally stored observation
+     * @return Observable<Observation>
+     */
+    fun createObservationFromLocal(observationEntity: ObservationEntity): Observable<Observation>{
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt:38]:
```suggestion
    fun createObservationFromLocal(observationEntity: ObservationEntity): Observable<Observation> {
```

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        val observation = AppDatabaseHelper.convert(observationEntity)
+
+        return AppDatabaseHelper.createObservableIO(Callable{
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt:41]:
```suggestion
        return AppDatabaseHelper.createObservableIO(Callable {
```

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            if (!NetworkUtils.isOnline()) throw Exception("Must be online to fetch observations")
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt:42]:
```suggestion
            if (!NetworkUtils.isOnline()) throw Exception("To retrieve observations, an internet connection is required.")
```

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    /**
+     * Get an Observation from Observation Uuid from the server
+     *
+     * @param uuid the UUID of the observation
+     * @return Observable<Observation>
+     */
+    fun getObservationByUuid(uuid: String): Observable<Observation> {
+        return AppDatabaseHelper.createObservableIO(Callable{
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt:61]:
```suggestion
        return AppDatabaseHelper.createObservableIO(Callable {
```

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+            if (!NetworkUtils.isOnline()) throw Exception("Must be online to fetch observations")
+
+            restApi.getObservationByUuid(uuid).execute().run{
+                if (isSuccessful && body() != null) {
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt:65]:
```suggestion
            if (!NetworkUtils.isOnline()) throw Exception("To retrieve observations, an internet connection is required.")

            restApi.getObservationByUuid(uuid).execute().run {
                if (isSuccessful && body() != null) {
```

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+    /**
+     * Get all observation resources of a patient from the server.
+     *
+     * @param encounterUuid the UUID of the encounter
+     * @return the resource list of observations
+     */
+    fun getAllObservationResourcesByEncounterUuid(encounterUuid: String): Observable<List<Resource>> {
+        return AppDatabaseHelper.createObservableIO(Callable {
+            if (!NetworkUtils.isOnline()) throw Exception("Must be online to fetch observations")
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ObservationRepository.kt:102]: same

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/dao/ObservationDAO.java [hunk 8]
```diff
@@ -32,6 +34,37 @@ public class ObservationDAO {
...
+     *
+     * @param obs the observation model
+     * @param encounterID the encounterId
+     *
+     * @return id (primary key)
+     */
+    public Long saveObservation(Observation obs, long encounterID) {
+        ObservationEntity observationEntity = AppDatabaseHelper.convert(obs, encounterID);
+        Long id = observationRoomDAO.addObservation(observationEntity);
+        return id;
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/dao/ObservationDAO.java:50]:
```suggestion
        return observationRoomDAO.addObservation(observationEntity);
```

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/dao/ObservationDAO.java [hunk 9]
```diff
@@ -32,6 +34,37 @@ public class ObservationDAO {
...
+     * @param obs the observation
+     * @param encounterId the observation (needed to verify that the observation already exists)
+     *
+     * @return count of updated values
+     */
+    public int updateObservation(Observation obs, long encounterId) {
+        ObservationEntity observationEntity = AppDatabaseHelper.convert(obs, encounterId);
+        observationEntity.setId(encounterId);
+        int count = observationRoomDAO.updateObservation(observationEntity);
+        return count;
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/dao/ObservationDAO.java:65]:
```suggestion
        return observationRoomDAO.updateObservation(observationEntity);
```

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/dao/ObservationDAO.java [hunk 10]
```diff
@@ -49,4 +82,30 @@ public List<Observation> findObservationByEncounterID(Long encounterID) {
...
+     * @return the list of primary keys
+     */
+    public List<Long> saveStandaloneObservations(List<Observation> observationList) {
+        List<StandaloneObservationEntity> standaloneObservationEntityList = new ArrayList<>();
+        for(Observation observation: observationList){
+            StandaloneObservationEntity standaloneObservationEntity = AppDatabaseHelper.convertToStandalone(observation);
+            standaloneObservationEntityList.add(standaloneObservationEntity);
+        }
+        List<Long> id = observationRoomDAO.addStandaloneObservationList(standaloneObservationEntityList);
+        return id;
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/dao/ObservationDAO.java:99]: type convert and return without storing it..

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/dao/ObservationRoomDAO.kt [hunk 11]
```diff
@@ -13,14 +13,9 @@
...
 package com.openmrs.android_sdk.library.dao
 
-import androidx.room.Dao
-import androidx.room.Delete
-import androidx.room.Insert
-import androidx.room.OnConflictStrategy
-import androidx.room.Query
-import androidx.room.Update
-
+import androidx.room.*
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/dao/ObservationRoomDAO.kt:16]: star imports

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/databases/AppDatabaseHelper.kt [hunk 12]
```diff
@@ -321,6 +353,25 @@ object AppDatabaseHelper {
         return allergy
     }
 
+    @JvmStatic
+    fun convertToStandalone(observation: Observation): StandaloneObservationEntity {
+        val standaloneObservationEntity = StandaloneObservationEntity(
```
Comment (rishabh-997) [inline openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/databases/AppDatabaseHelper.kt:358]:
```suggestion
        return StandaloneObservationEntity(
```

**Review by rishabh-997 — CHANGES_REQUESTED** [review]
Looks good

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/EncounterRepository.kt (hunk 1 above)
Comment (shubhamsgit) [reply openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/EncounterRepository.kt]: My bad :( Done!

**Comment by rishabh-997** [comment]
New changes look good. Thanks

**Pull request merged**